external-crates/move/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -140,84 +126,728 @@ struct PackageWithResolverTOML<'a>(&'a Package);
+                    let (other_pkg_deps, pkg_deps) = deps_equal(*dep_name, &other_g, &g);
+                    if other_pkg_deps != pkg_deps {
+                        bail!(
+                            "When resolving dependencies for package {}, \
+                             conflicting dependencies found:\n{}{}",
+                            root_package,
+                            format_deps(
+                                dep_path_from_root(
+                                    root_package,
+                                    &other_g,
+                                    *dep_name,
+                                    *is_other_external
+                                )?,
+                                other_pkg_deps
+                            ),
+                            format_deps(
+                                dep_path_from_root(root_package, &g, *dep_name, *is_external)?,
+                                pkg_deps
+                            ),
+                        )
+                    }
+                }
+            }
+
+            if internally_resolved {
+                // insert edges from the directly dependent package to its neighbors for
+                // internally resolved sub-graphs - due to how external graphs are constructed,
+                // edges between directly dependent packages and their neighbors are already in
+                // the sub-graph and would have been inserted in the first loop in this function
+                for to_pkg_name in g
+                    .package_graph
+                    .neighbors_directed(*dep_name, Direction::Outgoing)
+                {
+                    // unwrap is safe as all edges have a Dependency weight
+                    let sub_dep = g.package_graph.edge_weight(*dep_name, to_pkg_name).unwrap();
+                    self.package_graph
+                        .add_edge(*dep_name, to_pkg_name, sub_dep.clone());
+                }
+            }
+        }
+        Ok(())
     }
 
-    /// Create a dependency graph by reading a lock file.
+    /// Inserts a single direct dependency with given (package) name representing a sub-graph into
+    /// the combined graph. Returns true if the sub-graph has to be further merged into the combined
+    /// graph and false if it does not (i.e., if the dependency is already represented in the
+    /// combined graph).
+    fn insert_direct_dep(
+        &mut self,
+        dep: &PM::Dependency,
+        root_package: PM::PackageName,
+        dep_pkg_name: PM::PackageName,
+        sub_graph: &DependencyGraph,
+        mode: DependencyMode,
+        parent: &PM::DependencyKind,
+    ) -> Result<bool> {
+        match dep {
+            PM::Dependency::Internal(internal) => {
+                let PM::InternalDependency {
+                    kind,
+                    version,
+                    subst,
+                    digest,
+                    dep_override,
+                } = internal;
```

Comment:
Some unpacked fields are also used when creating `Dependency`

Reply:
Not sure I follow on this one, I'm suggesting that instead of

```rust
    PM::Dependency::Internal(internal) => {
        let PM::InternalDependency { 
            /* ... */
        } = internal;
```

You have:

```rust
    PM::Dependency::Internal(PM::InternalDependency { 
        /* ... */
    }) => {
```

Because I don't see any other reference to `internal` in the rest of the body.  I didn't think the fact that the unpacked fields are used to create a `Dependency` would make a difference here but maybe I'm missing something.

Reply:
Unpacking of `internal` defines a bunch of local variables (e.g., `subst`) that are used outside of the `match` pattern.

Reply:
Maybe I'm missing something super obvious, if so apologies, but to me, the body of this function looks like this:

```rust
fn insert_direct_dep(...) {
    match dep {
        Dependency::Internal(internal) => {
            let PM::InternalDependency {
                ... (1)
            } = internal;

            ... (2)
        }

        Dependency::External(_) => {
            ...
        }
    }
}
```

Where all the references to variables bound in (1) are only used in (2) (i.e. there's literally nowhere else you could refer to those bindings), so the suggestion to turn this structure into:

```rust
fn insert_direct_dep(...) {
    match dep {
        Dependency::Internal(PM::InternalDependency { ... (1) }) => {
            ... (2)
        }

        Dependency::External(_) => {
            ...
        }
    }
}
```

should be fine?

Again, entirely possible that I'm missing something obvious.

Reply:
I totally misunderstood what you were suggesting :-) and now that I do, I made the change.

For the record, there was another match in this function (now replaced with `if let Entry::Vacant(entry)...`) and I somehow thought you suggested moving struct unpack inside this match statement...